### ISSUE TITLE
Mastering Embedded Linux Part 1: Concepts

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6066]
Mastering Embedded Linux, Part 1: Concepts<br>
<br>
https://ift.tt/2PUJaDO<br>
<br>
<br>
<br>
<div>
<div>
<p>One of my favorite hobbies is hacking low-cost embedded systems that run Linux, or can be persuaded to. These systems are absolutely everywhere: the combination of powerful yet inexpensive processors, mass production of consumer goods, and the flexibility of Linux means that hobbyists can buy or build an embedded device capable of running Linux, often for less than $10.</p>
<p>I'll say that again: not only is embedded Linux within reach for makers and hobbyists, it's also <em>cheap</em> to throw into your next design.</p>
<p>I'd like to show you just how diverse this ecosystem is, and then I'd like to <strong>empower you to start tinkering with these systems</strong>. This article series will walk you through choosing hardware and compiling your first system image from source code. Then, we'll dive deeper into how to pick components for building a custom Linux board, how to deliver updates in a reliable, production-ready way, and how to customize the Linux kernel.</p>
<p>In short, this series of articles will get you up and running with the embedded Linux ecosystem. Ideally you should have some Linux command line knowledge, a good understanding of the overall components of an embedded system (flash memory, processor, peripherals), and a good chunk of free time for learning and experimenting.</p>
<figure><img src="https://imgs.xkcd.com/comics/cautionary.png"></figure>
<p>Let's start with some high-level concepts. Having a big-picture view will help prevent being overwhelmed when we start going into more detail later.</p>
<h1>Embedded Linux concepts</h1>
<p>If you're already familiar with microcontrollers, you're by-and-large used to having everything in a single package. You buy an STM32F1 or whatever, and you get flash program memory, RAM, a processor core or two, and some peripherals. If you buy a part specifically designed for it, you might get one or two really nice peripherals tailored to an application &mdash; built-in Bluetooth, USB 3.0, or a MIPI camera interface. The device probably has a development studio in which you write your code and send it over to run &ldquo;bare metal&rdquo; without a real operating system to get in its way. Finally, all this tends to come in a fairly hobbyist-friendly package: SOIC or QFP packages are usually available for at least one part in a microcontroller line.</p>
<p>graph BT; subgraph Microcontroller A[Processor core] --- B[RAM] A --- ST[Storage] subgraph Peripherals D[USB] E[ADC] S[SPI] U[UART] end A --- D A --- E A --- U A --- S end</p>
<p>The architecture is a little different with Linux-capable processors. Typically, these processors cram in nearly every peripheral you can think of, and many you haven't heard of. In exchange, they frequently lack built-in RAM, and always lack storage. On top of that, starting one of these devices up is a more involved setup: the processor typically has a built-in boot ROM that is responsible only for loading your bootloader.</p>
<p>graph BT; B[RAM] subgraph Microprocessor A[Processor core] subgraph Peripherals D[USB] E[ADC] S[SPI] U[UART] end A --- D A --- E A --- U A --- S BC[Boot ROM] --- A end B --- A ST[Storage] ST --- S</p>
<p>From there, the bootloader will load Linux, which will configure the rest of the peripherals and run a preset series of programs. Networking, application logic, and user interface are all available, just like a bigger Linux system. As far as the application is concerned, the environment looks identical to a desktop Linux system &mdash; all the APIs are the same, the filesystem looks normal, and it can talk to the outside world using normal networking protocols.</p>
<p>graph TB; Hardware --&gt; Bootloader Bootloader --&gt; Linux Linux --&gt; A1[SSH server] Linux --&gt; A2[Web server] Linux --&gt; A3[User interface] Linux --&gt; A4[Custom programs]</p>
<p>All this is stored in a tiny system image that can be as small as 4MB, depending on the needs of the device. But this is a key point, so I'll say it again: <strong>every software component running on Linux is nearly identical to the version you'd run on a desktop!</strong> Linux provides all the abstraction your applications need, so your job is mostly reduced to getting the bootloader and Linux kernel ported to your hardware.</p>
<h1>Components</h1>
<p>Almost every one of these components is replaceable. The choices made for each of these constitute a large portion of the design decisions you make while choosing an embedded system.</p>
<h2>Microprocessor</h2>
<p>Of course, your processor will determine a lot of your system's capabilities. Most of the small, cheap embedded Linux systems you'll be interacting with will either be using an ARM or MIPS core, and most of the industry is moving further and further toward ARM. (Keep your eye on <a href="https://riscv.org/">RISC-V</a>; it's an open-source instruction set architecture that isn't quite ready for prime time but has already seen a lot of industry interest.)</p>
<p>On top of the architecture, each silicon vendor adds many different peripherals. Much like their microcontroller brethren, these are small bits of silicon that do one thing, like USB or SPI. They are often configured via memory-mapped registers, and often, manufacturers will add more capabilites to a new product (a &ldquo;part&rdquo;) by copy-pasting peripheral IP from an older part, or by adding more copies of the peripheral to the new part.</p>
<h2>Memory</h2>
<p>These processors typically don't ship with memory. This is provided separately. The usual kinds of memory you might be familiar with apply here: SDRAM is often used on the low end, while DDR, DDR2, and DDR3 are used for more powerful parts. Although I haven't yet seen a whole lot of devices using DDR4, I'm sure we'll get there. The processor provides a module to manage the memory; this module is either initialized by the manufacturer's boot code or by the first stage of the bootloader.</p>
<p>Note how I said processors <em>typically</em> don't ship with RAM embedded in their package? Sometimes, they do. If you're buying an embedded system on a module, or a ready-to-go development board, you don't care about this, because someone else has already done the hard work of putting the memory on the PCB.</p>
<p>But if you <em>are</em> a hobbyist building your own from individual parts, you care very deeply about these parts (whether or not you know it yet). That's because laying memory out on a PCB is difficult for a few different reasons. So when I'm talking about hardware in my next article, I'll be sure call out some parts I've found that ship their own memory.</p>
<h2>Storage</h2>
<p><em>Storage</em>, distinct from memory, is the involatile place your code and data is stored. On embedded systems, it will almost certainly be some form of flash memory. Even in cases where the system boots from the network, typically this is still assisted by a bootloader stored in flash.</p>
<p>There are multiple kinds of flash memory. You are likely already familiar with SD or microSD cards. These provide raw storage, plus a <em>flash controller</em> whose job is to abstract the raw flash into a well-behaved storage medium (more on this in a second). SD cards are not known for their reliability, as anybody who has used a Raspberry Pi can tell you. However, later in this series, I'll explain how best to work around questionably reliable storage media.</p>
<p><em>eMMC</em> is an embeddable version of an SD card. Conceptually it is very similar; it has raw flash and a controller bundled into a single part. Unfortunately it is usually sold in nasty <a href="https://en.wikipedia.org/wiki/Ball_grid_array">BGA packages</a> like is pictured, which aren't a problem for large companies but are pretty much unusable for hobbyists because you can't easily solder it.</p>
<figure><img src="https://www.thirtythreeforty.net/media/embedded-linux-emmc.jpg">
<figcaption>
<p><a href="https://commons.wikimedia.org/wiki/File:EMMC.jpg">via Wikimedia Foundation</a></p>
</figcaption>
</figure>
<p>You can also buy raw flash without a controller attached to it. This is typically cheaper than eMMC and is sold in two varieties, NOR and NAND. <em>NOR flash</em> is slow to write, very cheap, and not very dense. Boot ROMs almost always know how to read from NOR flash accessed over SPI. If you're not writing very often, maybe only once to burn firmware to the device, it's a great option. <em>NAND flash</em> is faster, denser, and a little more expensive. You can buy NAND that's accessed via SPI, but there are also faster parts that can be accessed over a dedicated NAND bus.</p>
<p>Raw flash isn't really all that nice. You cannot write to it willy-nilly; flash can only be written once before you must erase it in large blocks. Furthermore, flash can be written a limited number of times before it wears out (between 1000 and 100000, depending on the technology).</p>
<p>A software guy's first instinct is to throw software at a hardware problem, and that's what we'll do here, later on. A Linux subsystem called UBI can help work around these limitations and make raw flash nicer to use.</p>
<h1>Software</h1>
<p>If you're developing for one of these systems, you provide the bootloader, kernel, and filesystem. Thankfully, once Linux has booted, it often supports many of the peripherals on a system out of the box using a wonderfully consistent set of driver interfaces &mdash; in stark contrast to how often you must write drivers yourself on a microcontroller.</p>
<p>In general the software typically follows a very understandable pattern that you can apply anywhere.</p>
<p>You don't have to micromanage all these pieces at once. There are embedded Linux distributions that provide a complete toolkit to help you build a firmware image with all of this included. Later on, we'll start using a distribution and compile our own complete system.</p>
<h2>Bootloader</h2>
<p>The bootloader is the first program that the engineer or hacker has control over. It is only as complex as is needed to load the kernel and get it running. In practice, the bootloader can still be pretty complex.</p>
<p>Your embedded Linux system will almost certainly be using <a href="https://en.wikipedia.org/wiki/Das_U-Boot">Das U-Boot</a>, the so-called &ldquo;universal bootloader.&rdquo; (It really does run on nearly everything!) The bootloader has stripped-down drivers for the onboard storage, perhaps a couple of other peripherals, and just enough code to read the kernel into memory and start executing it.</p>
<p>A lot of times, people don't mess with the bootloader that comes with their board, and they just follow the conventions it's expecting. That's fine, but often you want your system to do something the stock bootloader can't. With the right tools, you have no reason to be afraid of modifying the bootloader &mdash; it is a program like any other.</p>
<h3>Boot ROM</h3>
<p>I should mention the Boot ROM here. The Boot ROM is a small chunk of code embedded in the processor, provided by the manufacturer. It is very, very simple, and typically it immediately runs the real bootloader from a couple different storage media &mdash; often this &ldquo;boot order&rdquo; is specified in the processor's datasheet.</p>
<p>Boot ROMs have one other cool trick &mdash; they often speak USB! This allows you to connect a brand-new, unprogrammed system to a computer over USB and run code, or flash a storage medium. If it's present, this feature makes a board nearly impossible to brick. Different manufacturers call this mode different things. NXP/Freescale calls it <a href="https://github.com/NXPmicro/mfgtools/wiki">Download Mode</a>. Allwinner calls it <a href="https://linux-sunxi.org/FEL">FEL mode</a>. You will generally need a special program on your workstation to use this mode, and capabilities vary between different processors.</p>
<h2>Kernel</h2>
<p>You already know what Linux is, I hope! Linux must be <em>ported</em> to each architecture, each part, and each board. All these drivers ship with the kernel source. There's a lot of them, but because of the tendency of manufacturers to reuse IP, it's managable. Typically, a port consists of the following:</p>
<ul>
<li>
<strong>Architecture code</strong> provides low-level routines for very basic things like register manipulation, synchronization primitives, etc. Porting Linux to a completely new architecture is an immense amount of work, and you will likely not have to do this anytime soon.</li>
<li>
<strong>Drivers</strong> make up the bulk of the kernel source code. This is because Linux ships drivers for every device supported by Linux, in one source code tree. Most of them are not needed by your embedded system &mdash; for example, a MIPS router has absolutely no use for a driver for Intel QuickSync. So most of these drivers are not even compiled for very small systems.</li>
<li>
<strong>The device tree</strong> is a very important part of the port that explains to the kernel how hardware is actually connected to the system. Device trees are the &ldquo;config files&rdquo; for Linux drivers. This means that Linux drivers are easy to reuse &mdash; if you add a device tree entry for a piece of hardware on your system and compile the relevant driver, the driver will find the entry and set itself up. We'll talk a lot about device trees later.</li>
</ul>
<h2>Userland</h2>
<p>Like I mentioned already, the &ldquo;rest&rdquo; of the software on an embedded system is pretty much identical to its desktop counterparts. Together, this &ldquo;everything else&rdquo; is frequently called &ldquo;user land,&rdquo; because it's the land where user software can roam freely, without worries of the nasty hardware gremlins that lurk underneath the peaceful abstraction of the Linux API.</p>
<p>We'll go over userland in more detail later. The major components you need to know right now are the filesystem, the init system, and the shell.</p>
<p>The filesystem is important because some are better for embedded systems than others &mdash; remember the SD card reliability we talked about? Filesystems can help with this. If the system is using eMMC or an SD card, it looks like a normal &ldquo;<a href="https://unix.stackexchange.com/a/259200/31658">block device</a>,&rdquo; and you can use the standard ext2/3/4 or stuff like the <a href="https://en.wikipedia.org/wiki/F2FS">flash-friendly filesystem (f2fs)</a>, which helps with reliability in various ways. If you're on raw flash, you need to use more esoteric filesystems that are designed for raw flash, such as <a href="https://en.wikipedia.org/wiki/JFFS2">JFFS2</a> or <a href="https://en.wikipedia.org/wiki/UBIFS">UBIFS</a>. My personal favorite is to use <a href="https://en.wikipedia.org/wiki/SquashFS">squashfs</a> on an UBI partition; we'll talk more about this when we start customizing our firmware image!</p>
<p>Next, the init system is responsible for managing userland. You might have heard of systemd? It's used on bigger embedded systems. However, it's too big for the really small systems, which usually use a SysV init scheme of simple shell scripts. Most embedded Linux distributions provide this, and again, when we're customizing firmware, I'll show you how to add your programs to run automatically.</p>
<p>And finally, the <a href="https://en.wikipedia.org/wiki/Bash_(Unix_shell)">shell</a> is what you'll interact with. Traditionally, this is done over a UART serial connection, but occasionally you might have the luxury of a screen. If you get your system to a shell prompt, you have definitely gotten your system up and running!</p>
<h1>Next steps</h1>
<p>Okay. That's a lot of info all at once. Thanks for staying with me!</p>
<p>With that out of the way, here's my rough plan for the series:</p>
<ul>
<li>
<strong>Hardware</strong>: We'll choose a development board. Spoilers: I'll use a Raspberry Pi Zero for demo for the first few software tutorials. But here, I'll also point out various other cheap Linux-capable hardware. If you're interested in building your own, I am also recommending hacker-friendly parts that you can add to your own PCB.</li>
<li>
<strong>Buildroot</strong>: For a quick win, we'll download Buildroot and compile a complete operating system from scratch. We'll flash it and boot it on real hardware. I'll be sure to explain what's happening at each step.</li>
<li>
<strong>Customizing the firmware image</strong>: here we'll depart from what the Buildroot developers&rsquo; defaults and begin to make changes to the Raspberry Pi's firmware. Each individual step is pretty straightforward, and all combined, your finished system can look completely different from the stock one. It's all up to the engineer.</li>
<li>
<strong>Going smaller</strong>: We leave the sunny skies of Raspberry Pi land and start working on an image for a very small board with just 4MB of storage. It's quite impressive how much functionality you can cram in&mdash;<a href="https://i.redd.it/tfugj4n3l6ez.png">take that, node_modules</a>!</li>
<li>
<strong>Hacking on U-Boot and Linux</strong>: The available code for your board's processor is out of date. What do you do?</li>
<li>
<strong>Building a board from scratch</strong>: We'll go back to the parts I recommend in the Hardware stage and start building our very own dev board. (This part is subject to change as I figure out the best approach!)</li>
</ul>
<p>Next we'll look around at some hardware. If you'd like to follow along, you can grab my RSS feed from the sidebar.</p>
<p>In the meantime, if you'd like to Linux-enable your project, product, or program, leave a comment or <a href="mailto:thirtythreeforty@gmail.com">drop me a line</a>&mdash;I'd love to hear from you and I'm always happy to hear about cool stuff.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via<br>
<br>
December 27, 2019 at 12:07PM
